pkg/auth/azure.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019-present, Cruise LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package auth
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	cfg "github.com/cruise-automation/daytona/pkg/config"
+	"github.com/hashicorp/vault/api"
+)
+
+// AzureService is an external service that vault can authenticate request against
+type AzureService struct{}
+
+type azureVaultPayload struct {
+	Role              string
+	SubscriptionID    string
+	VmName            string
+	ResourceGroupName string
+	JWT               string
+}
+
+func (a *AzureService) Auth(client *api.Client, config cfg.Config) (string, error) {
+	metadata, err := a.getMetadata()
+	if err != nil {
+		return "", err
+	}
+
+	jwt, err := a.getJWT()
+	if err != nil {
+		return "", err
+	}
+
+	loginData := map[string]interface{}{
+		"role":                config.VaultAuthRoleName,
+		"subscription_id":     metadata.Compute.SubscriptionID,
+		"vm_name":             metadata.Compute.Name,
+		"resource_group_name": metadata.Compute.ResourceGroupName,
+		"jwt":                 jwt,
+	}
+
+	return fetchVaultToken(client, config, loginData)
+}
+
+type simplifiedToken struct {
+	AccessToken string `json:"access_token"`
+}
+
+func (a *AzureService) getJWT() (string, error) {
```

Comment:
Minor: The `get` functions have a decent amount of duplication, is there any way to reduce the http request code?

Reply:
combined the common logic into a single func